polkadot/xcm/xcm-executor/src/lib.rs

```diff
@@ -469,31 +473,112 @@ impl<Config: config::Config> XcmExecutor<Config> {
 		Ok(())
 	}
 
-	fn take_fee(&mut self, fee: Assets, reason: FeeReason) -> XcmResult {
+	fn take_fee(&mut self, fees: Assets, reason: FeeReason) -> XcmResult {
 		if Config::FeeManager::is_waived(self.origin_ref(), reason.clone()) {
 			return Ok(())
 		}
 		tracing::trace!(
 			target: "xcm::fees",
-			?fee,
+			?fees,
 			origin_ref = ?self.origin_ref(),
 			fees_mode = ?self.fees_mode,
 			?reason,
 			"Taking fees",
 		);
-		let paid = if self.fees_mode.jit_withdraw {
+		// We only ever use the first asset from `fees`.
+		let asset_needed_for_fees = match fees.get(0) {
+			Some(fee) => fee,
+			None => return Ok(()), // No delivery fees need to be paid.
+		};
+		// If `BuyExecution` was called, we use that asset for delivery fees as well.
+		let asset_to_pay_for_fees =
+			self.calculate_asset_for_delivery_fees(asset_needed_for_fees.clone());
+		tracing::trace!(target: "xcm::fees", ?asset_to_pay_for_fees);
+		// We withdraw or take from holding the asset the user wants to use for fee payment.
+		let withdrawn_fee_asset = if self.fees_mode.jit_withdraw {
 			let origin = self.origin_ref().ok_or(XcmError::BadOrigin)?;
-			for asset in fee.inner() {
-				Config::AssetTransactor::withdraw_asset(&asset, origin, Some(&self.context))?;
-			}
-			fee
+			Config::AssetTransactor::withdraw_asset(
+				&asset_to_pay_for_fees,
+				origin,
+				Some(&self.context),
+			)?;
+			tracing::trace!(target: "xcm::fees", ?asset_needed_for_fees);
+			asset_to_pay_for_fees.clone().into()
+		} else {
+			let assets_taken_from_holding_to_pay_delivery_fees = self
+				.holding
+				.try_take(asset_to_pay_for_fees.clone().into())
+				.map_err(|_| XcmError::NotHoldingFees)?;
+			tracing::trace!(target: "xcm::fees", ?assets_taken_from_holding_to_pay_delivery_fees);
+			let mut iter = assets_taken_from_holding_to_pay_delivery_fees.fungible_assets_iter();
+			let asset = iter.next().ok_or(XcmError::NotHoldingFees)?;
+			asset.into()
+		};
+		// We perform the swap, if needed, to pay fees.
+		let paid = if asset_to_pay_for_fees.id != asset_needed_for_fees.id {
+			let swapped_asset: Assets = Config::AssetExchanger::exchange_asset(
```

Comment:
we need to update the doc for `Config::AssetExchanger`, and tell that we use only first asset

Reply:
We only use the first asset for the purposes of swapping for delivery fees. `Config::AssetExchanger` still is able to take more than one asset. This is possible within the `ExchangeAsset` instruction, for example.

That said, I'm fine adding extra documentation about the fact that we only use the first asset for swapping anywhere you deem relevant. For now it's just in the prdoc and in the executor. I should add a comment in the router as well.

Reply:
I think it would be useful to mention this case where it is used and how.